internal/settings/linode.go

```diff
@@ -0,0 +1,321 @@
+	netlib "github.com/qdm12/golibs/network"
+)
+
+type linode struct {
+	domain    string
+	host      string
+	ipVersion models.IPVersion
+	token     string
+}
+
+func NewLinode(data json.RawMessage, domain, host string, ipVersion models.IPVersion,
+	_ bool, _ regex.Matcher) (s Settings, err error) {
+	extraSettings := struct {
+		Token string `json:"token"`
+	}{}
+	if err := json.Unmarshal(data, &extraSettings); err != nil {
+		return nil, err
+	}
+	l := &linode{
+		domain:    domain,
+		host:      host,
+		ipVersion: ipVersion,
+		token:     extraSettings.Token,
+	}
+	if err := l.isValid(); err != nil {
+		return nil, err
+	}
+	return l, nil
+}
+
+func (l *linode) isValid() error {
+	if len(l.token) == 0 {
+		return ErrEmptyToken
+	}
+	return nil
+}
+
+func (l *linode) String() string {
+	return toString(l.domain, l.host, constants.LUADNS, l.ipVersion)
+}
+
+func (l *linode) Domain() string {
+	return l.domain
+}
+
+func (l *linode) Host() string {
+	return l.host
+}
+
+func (l *linode) DNSLookup() bool {
+	return true
+}
+
+func (l *linode) IPVersion() models.IPVersion {
+	return l.ipVersion
+}
+
+func (l *linode) BuildDomainName() string {
+	return buildDomainName(l.host, l.domain)
+}
+
+func (l *linode) HTML() models.HTMLRow {
+	return models.HTMLRow{
+		Domain:    models.HTML(fmt.Sprintf("<a href=\"http://%s\">%s</a>", l.BuildDomainName(), l.BuildDomainName())),
+		Host:      models.HTML(l.Host()),
+		Provider:  "<a href=\"https://cloud.linode.com/\">Linode</a>",
+		IPVersion: models.HTML(l.ipVersion),
+	}
+}
+
+// Using https://www.linode.com/docs/api/domains/
+func (l *linode) Update(ctx context.Context, client netlib.Client, ip net.IP) (newIP net.IP, err error) {
+	domainID, err := l.getDomainID(ctx, client)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrGetDomainID, err)
+	}
+
+	recordType := A
+	if ip.To4() == nil {
+		recordType = AAAA
+	}
+
+	recordID, err := l.getRecordID(ctx, client, domainID, recordType)
+	if errors.Is(err, ErrNotFound) {
+		err := l.createRecord(ctx, client, domainID, recordType, ip)
+		if err != nil {
+			return nil, fmt.Errorf("%w: %s", ErrCreateRecord, err)
+		}
+		return ip, nil
+	} else if err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrGetRecordID, err)
+	}
+
+	if err := l.updateRecord(ctx, client, domainID, recordID, ip); err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrUpdateRecord, err)
+	}
+
+	return ip, nil
+}
+
+type linodeError struct {
+	Field  string `json:"field"`
+	Reason string `json:"reason"`
+}
+
+func (l *linode) getDomainID(ctx context.Context, client netlib.Client) (domainID int, err error) {
+	u := url.URL{
+		Scheme: "https",
+		Host:   "api.linode.com",
+		Path:   "/v4/domains",
+	}
+
+	r, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
+	if err != nil {
+		return 0, err
+	}
+	r.Header.Set("Content-Type", "application/json")
+	r.Header.Set("Authorization", "Bearer "+l.token)
+	r.Header.Set("oauth", "domains:read_only")
```

Comment:
I may be mistaken, but given we're using a personal access token, I don't think we need to define oauth headers?

Reply:
Hey thanks for the review 👍 

That's what they have on their documentation.
You can probably define tokens with different kind of permissions.
It might not work if you don't precise which permission you want, but I can't test it for sure, I don't have an account 😉

Reply:
Sounds good! When creating the token, I had to assign permissions to it, but better to go with what the documentation says :)

Reply:
Oh interesting which permissions? Are these the same as in the code? So I can add these to the Linode documentation.

Reply:
Go to setting and API Tokens:
<img width="414" alt="Screen Shot 2021-01-17 at 6 56 51 PM" src="https://user-images.githubusercontent.com/1011628/104867637-2f147100-58f6-11eb-9cb8-5aba7c439a38.png">

API Tokens Tab:
<img width="861" alt="Screen Shot 2021-01-17 at 6 57 38 PM" src="https://user-images.githubusercontent.com/1011628/104867652-389dd900-58f6-11eb-9301-1c0839554cc9.png">

Add Personal Access Token.
<img width="474" alt="Screen Shot 2021-01-17 at 6 53 54 PM" src="https://user-images.githubusercontent.com/1011628/104867665-405d7d80-58f6-11eb-93dd-38af9d40b5d1.png">

I have a token with `domains` set, with read and write privileges, set to never expire